ote_cli/ote_cli/utils/hpo.py

```diff
@@ -0,0 +1,660 @@
+            full_dataset_size=train_dataset_size,
+            expected_time_ratio=expected_time_ratio,
+            non_pure_train_ratio=val_dataset_size
+            / (train_dataset_size + val_dataset_size),
+            batch_size_name="learning_parameters.batch_size",
+        )
+
+        if self.algo == "smbo":
+            hpopt_arguments["num_init_trials"] = hpopt_cfg.get("num_init_trials")
+            hpopt_arguments["num_trials"] = hpopt_cfg.get("num_trials")
+        elif self.algo == "asha":
+            hpopt_arguments["num_brackets"] = hpopt_cfg.get("num_brackets")
+            hpopt_arguments["min_iterations"] = hpopt_cfg.get("min_iterations")
+            hpopt_arguments["reduction_factor"] = hpopt_cfg.get("reduction_factor")
+            hpopt_arguments["num_trials"] = hpopt_cfg.get("num_trials")
+            hpopt_arguments["num_workers"] = (
+                num_available_gpus // self.num_gpus_per_trial
+            )
+
+        HpoManager.remove_empty_keys(hpopt_arguments)
+
+        self.hpo = hpopt.create(**hpopt_arguments)
+
+    def check_resumable(self):
+        """
+        Check if HPO could be resumed from the previous result.
+        If previous results are found, ask the user if resume or start from scratch.
+        """
+        resume_flag = False
+
+        hpo_previous_status = hpopt.get_previous_status(self.work_dir)
+
+        if hpo_previous_status in [
+            hpopt.Status.PARTIALRESULT,
+            hpopt.Status.COMPLETERESULT,
+        ]:
+            print(f"Previous HPO results are found in {self.work_dir}.")
+            retry_count = 0
+
+            while True:
+                if hpo_previous_status == hpopt.Status.PARTIALRESULT:
+                    user_input = input(
+                        "Do you want to resume HPO? [\033[1mY\033[0m/n]: "
+                    )
+                else:
+                    user_input = input(
+                        "Do you want to reuse previously found "
+                        "hyper-parameters? [\033[1mY\033[0m/n]: "
+                    )
+
+                input_len = len(user_input)
+
+                if input_len < 1:
+                    # It means that an user accepted the default choice.
+                    resume_flag = True
+                    break
+
+                if input_len == 1:
+                    user_input = user_input.lower()
+                    if user_input in ["y", "n"]:
+                        if user_input == "y":
+                            resume_flag = True
+                        break
+
+                retry_count += 1
+                if retry_count >= 5:
+                    print(
+                        "Your inputs are invalid. "
+                        "The whole program is terminating..."
+                    )
+                    sys.exit()
+
+                print("You should type 'y' or 'n'. Nothing means 'y'.")
+        return resume_flag
+
+    def run(self):
```

Comment:
This function seems quite complicated. Can it be refactored and maybe split on several parts. Also time.sleep is not best way for synchronising processes. Could you consider using something like this? https://stackoverflow.com/questions/30686295/how-do-i-run-multiple-subprocesses-in-parallel-and-wait-for-them-to-finish-in-py
I would see several stages inside this function:
1. Prepare parameters
2. Run processes and wait for completion 
3. Gather results and determine which parameters are the best

Reply:
When using HPO ASHA, it is difficult to use blocking functions such as wait() because it is not possible to predict in advance which of the concurrently running processes will end first. However, considering one of the three cases of using time.sleep(), a blocking function can be used, so that part has been modified. (https://github.com/openvinotoolkit/training_extensions/pull/799#discussion_r775447500)